recipes/onnxruntime/all/patches/1.15.1-0002-cuda-gsl.patch

```diff
@@ -0,0 +1,12 @@
+diff --git a/cmake/external/onnxruntime_external_deps.cmake b/cmake/external/onnxruntime_external_deps.cmake
+index 9effd1a2db..3284c35424 100644
+--- a/cmake/external/onnxruntime_external_deps.cmake
++++ b/cmake/external/onnxruntime_external_deps.cmake
+@@ -288,6 +288,7 @@ if(onnxruntime_USE_CUDA)
+     URL ${DEP_URL_microsoft_gsl}
+     URL_HASH SHA1=${DEP_SHA1_microsoft_gsl}
+     PATCH_COMMAND ${Patch_EXECUTABLE} --binary --ignore-whitespace -p1 < ${PROJECT_SOURCE_DIR}/patches/gsl/1064.patch
++    FIND_PACKAGE_ARGS 4.0 NAMES Microsoft.GSL
```

Comment:
Can you submit this patch upstream?

Reply:
Thanks. I'm starting to wonder if we can go even further upstream. Would it be possible for the GSL to be updated with this path to support CUDA? That would be the ideal place to fix this.

Reply:
Possibly, I did not look at the specific patch onnxruntime applies. I do think it is up to onnxruntime to take this up and consider it out of scope for here. It would be a transparent change for this conan package.

Reply:
Well, I'm concerned that the patch is applied to the GSL Conan package in the Conan cache. Modifying a package in the cache in this way could cause unexpected problems since the package won't correlate to its hash after the patch is applied.

Reply:
It will patch the version downloaded using Git somewhere below the build folder of the onnxruntime package. It shouldn't touch Conan's GSL package directory.

Since I get an error during CMake config, not during the build, it implies that the GSL override from the Conan GSL package is probably not working.

Reply:
@bverhagen, looking into this, it appears that adding `FIND_PACKAGE_ARGS` opts in to using `find_package` to attempt to find the installed version. Refer to the section on `FIND_PACKAGE_ARGS` [here](https://cmake.org/cmake/help/latest/module/FetchContent.html). Since Conan requires ms-gsl, this should find the Conan installed version. It would then end up patching it. I think this would need to be fixed upstream in Microsoft.GSL in order to use the ms-gsl Conan package when CUDA is enabled.

Reply:
Per the docs you linked: FIND_PACKAGE_ARGS is trying to resolve the package from the system using find_package() with the arguments you give it. If it finds it, it uses it and does not call `ExternalProject` with the remainder of the options (including PATCH_COMMAND).

Hence when it finds the system or conan `ms-gsl`, it is not going to patch it but simply use it instead (I guess the patch is no longer necessary or conan sets the right things).

It would also be pretty insane for CMake trying to patch it, as find_package() is intended to find system packages, which are typically not writable. Even more insane it would be to rewrite these for the whole system.

The patch is actually already merged upstream (see https://github.com/microsoft/GSL/pull/1064), but not included in 4.0.0.

Even though we don't seem to have problems using conan's ms-gsl 4.0.0, it seems advisable to apply the patch from onnxruntime until a new version of ms-gsl is out. I'll adapt the requirements and add a summary of the above as a comment.

Reply:
@jwillikers : I tried changing the PR so it does not depend on Conan's ms-gsl when CUDA is enabled. However, I got trapped in a situation where FetchContent_Declare in onnxruntime cmake does not download gsl when called from Conan while it works perfectly when called directly from the CLI.

Since it is only a problem until the next GSL release, I decided to not dive in further but to go for the second best option and use Conan's ms-gsl in all cases. To resolve your worries, I'll delete the `PATCH_COMMAND` in the *-cuda-gsl.patch files. Is that a good enough solution for you?

Reply:
I created a follow-up PR #20619 that gets rid of the FetchContent calls entirely. Could you test that one with `use_cuda=True`, perhaps?